### DISCUSSION TITLE
How to run commands in the project root directory?

### DISCUSSION THREAD

**Discussion by abersheeran** [pdm-project/pdm#470] [Q&A]
pdm can run commands that defined in pyproject.toml in a certain directory in the project, but the value of cwd is the path to run `pdm run ...`. Is there a solution (such as pyproject.toml configuration) to ensure the working directory of the command always in the project root directory?

**Comment by frostming** [comment]
No, there isn't such configuration. In what case you run the script in a sub path while still wanting the cwd to be the project root?

Configure the cwd to run script is just like you cd to that directory and run, so this doesn't make sense.

**Reply by abersheeran** [reply]
Yes. How should I do?

**Reply by frostming** [reply]
I am raising a question, not an answer.

Imagine you have a script under `project/deep/path/my_script.py`, you just `python deep/path/my_script.py`, or `python -m deep.path.my_script` to ensure the package path won't change

**Reply by abersheeran** [reply]
In this case, I always use `Path(__file__).parent` to ensure cwd in project root.

**Comment by pawamoy** [comment]
Does PDM sets the `PROJECT_ROOT` variable in the environment? Because if it does, one could use that variable in scripts to make sure they run in the project root. (I might have misunderstood the question and replies though)

**Reply by frostming** [reply]
No, it is a magic variable, not set in env var.

**Reply by pawamoy** [reply]
It could be interesting to add it to the env then, at least for the `run` command.
This way, whatever the subdir I'm in, I could use `pdm run some-tool some-arg` as if I was in the root (given `some-tool` uses `PROJECT_ROOT` to cd into it). That would play well with task runners such as invoke or duty ^^

**Reply by abersheeran** [reply]
> This way, whatever the subdir I'm in, I could use `pdm run some-tool some-arg` as if I was in the root (given `some-tool` uses `PROJECT_ROOT` to cd into it). That would play well with task runners such as invoke or duty ^^

I agree you.

**Comment by frostming** [comment]
It is done in 2ce445176a68b7696ce3cabb4ad288dc8d0e3851

**Reply by abersheeran** [reply]
I don’t understand too much. After the new version is released, how can I use it?

**Reply by frostming** [reply]
> given some-tool uses `PROJECT_ROOT` to cd into it

**Reply by abersheeran** [reply]
Just like `cd $PROJECT_ROOT && some-command`?

**Reply by frostming** [reply]
Oh, there must be scenarios that users want to run the scripts in other paths so that relative paths are calculated as they expect.

**Reply by abersheeran** [reply]
I don't quite understand, can you give me an example? For example, my Dockerfile is in the root directory, but I want to execute the packaged `pdm run build-docker` in a subdirectory.

**Reply by frostming** [reply]
```toml
analyze = "python -m code.analyze"
```

And the code analysis is done against the code in the current working directory instead of root directory. ~~If your script must be done against root directory, you should be able to cd to PDM_PROJECT_ROOT in the script yourself.~~

**Reply by abersheeran** [reply]
The follow code is right?
```toml
# in pyproject.toml
build-docker = {shell = "cd $PDM_PROJECT_ROOT && docker build . -t name"}
```

**Reply by frostming** [reply]
I just tried yarn and it goes to project root, I am going to follow it, too.

**Reply by 164747** [reply]
This has been very confusing for me. I want to be able to run loads of files with the pdm intepreter that resides deep within my project. Test script etc, that depend on the entire interpreter and its dependancies. Now that is impossible. I would expect the project root always to be in sys.path, if it where, everything would work as expected, now I cannot run any files not in the root.

**Reply by pawamoy** [reply]
Some workarounds:

```bash
pdmroot() {
    pdm run bash <<< "cd '$PDM_PROJECT_ROOT; $*"
}  # probably not very robust
```

```bash
gitroot() (  # note the parenthesis: it runs in a subshell
    cd "$(git rev-parse --show-toplevel)"
    "$@"
)  # does not allow shell construct either (like >, &&, etc.)
```

**Reply by pawamoy** [reply]
> I just tried yarn and it goes to project root, I am going to follow it, too.

So you were planning to make PDM always cd into the project root? Or did I misunderstand?

**Reply by 164747** [reply]
If I set 

`export PYTHONPATH=$PYTHONPATH:./`

in the command terminal, I can run commands with

`pdm run python path/to/deep/file.py`

from my pdm root directory. Don't know if that's the right approach though.

**Reply by 164747** [reply]
Background, I have a project, a web service I deploy to google app engine For the main file `app.py` it is not problem to have that in the "pdm root". But I have a lot of "test script" that I like to run whilst developing. I don't want to bloat my "root" directory, so I want to call them like in my example above. 

`pdm run python path/to/deep/file.py`

Before I did the "export trick" above, all the import statements in the deep file stoped working.

**Reply by pawamoy** [reply]
Do you use relative imports? With proper packages structure and absolute imports it should work normally, whatever the directory you're currently in. ~~By the way, the current working directory is always added to sys.path by Python itself, so you shouldn't need to manually add it to PYTHONPATH.~~ Hmm in fact I believe Python adds the parent directory of the script you run, not the current working directory, so that explains your situation yes.

**Reply by frostming** [reply]
Basically when you run `python path/to/deep/file.py`, the directory `path/to/deep`, not project root, will be added to `sys.path`. This is the basic of Python import system. `pdm run` only adds `${project_root}/__pypackages__/3.10/lib` to the `sys.path`. If you want to refer to the modules under the project root, install it into the `__pypackages__`. This can be done by PDM automatically if your project has both `name` and `version` properties in metadata.
